In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df=pd.read_csv("../Resources/lending_data.csv")

# Review the DataFrame
lending_data_df

loan_size  interest_rate  borrower_income  debt_to_income   
0        10700.0          7.672            52800        0.431818  \
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df["loan_status"]

# Separate the X variable, the features
X = lending_data_df.drop(columns=["loan_status"])

In [4]:
# Review the y variable Series
print(y)

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64


In [5]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts   
0    10700.0          7.672            52800        0.431818                5  \
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [6]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [12]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using training data
classifier.fit(X_train,y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [13]:
# Make a prediction using the testing data
predictions = classifier.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [14]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predictions)


array([[18673,    86],
       [   32,   593]])

In [16]:
# Print the classification report for the model
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.95      0.91       625

    accuracy                           0.99     19384
   macro avg       0.94      0.97      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
- **precision:** The model achieved a perfect 100% precision for healthy loans, while high-risk loans had an 87% precision, meaning 13% were false positives.
- **recall:** It correctly identified 100% of healthy loans and 95% of high-risk loans, leaving 5% as false negatives.
- **f1-score:** The F1-score of 1 for healthy loans indicates an excellent balance between precision and recall, while 0.91 for high-risk loans demonstrates a strong but slightly less balanced performance.
- **accuracy:** The model correctly classified 99% of all samples.

Overall, this model performs well, demonstrating high precision and recall for both loan categories. With a 99% accuracy rate, it is highly reliable, though there is room for improvement, particularly in reducing the 13% false positive rate for high-risk loans.

## Create and Build Support Vector Machine Model

###  Step 1: Fit a Support Vector Machine Model by using the training data (`X_train` and `y_train`).

In [20]:
# Support Vector Machine classifier
from sklearn.svm import SVC
model = SVC(kernel='linear')
model.fit(X_train, y_train)


SVC(kernel='linear')

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [21]:
# Save the predictions
predictions = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:
* Print the classification report

In [22]:
# Calculate the classification report
print(classification_report(y_test, predictions))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.98      0.92       625

    accuracy                           0.99     19384
   macro avg       0.94      0.99      0.96     19384
weighted avg       1.00      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the support vector model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:**
The SVM model produced results similar to Logistic Regression, with few key differences:
- **Higher recall (98%),** meaning it reduced the number of false negatives compared to Logistic Regression.
- **Higher f1-score (92%),** indicating a better balance between precision and recall.
- **Higher macro-average score (99%)** indicating a better average score for high risk loans for precision, recall, and f1-score.

## Summary Analysis and Model Recommendation

Both the **Logistic Regression** and **Support Vector Machine (SVM)** models performed well in classifying loan risk, achieving high accuracy, precision, recall, and f1-scores. However, key differences distinguish their performance:

1. **Accuracy** – Both models correctly classified **99% of total samples**, indicating strong reliability.
2. **Precision** – Logistic Regression had a **100% precision for healthy loans** but an **87% precision for high-risk loans**, meaning 13% of high-risk loans were misclassified as low risk (false positives).
3. **Recall** – SVM outperformed Logistic Regression in recall for high-risk loans **(98% vs. 95%)**, reducing false negatives and ensuring more high-risk loans were correctly identified.
4. **Balance of Metrics** – SVM had a higher macro-average score (96%), indicating a better average of precision, recall, and F1-score across both loan categories.

### Recommended Model: Support Vector Machine (SVM)
The SVM model is recommended as the better option due to its higher recall and better overall balance in classifying both healthy and high-risk loans. SVM’s higher recall rate reduces the risk of falsely classifying high-risk loans as low risk**—a critical factor in financial decision-making**. The improved macro-average score also suggests that SVM maintains a stronger equilibrium between all evaluation metrics.

For real-world loan assessments, reducing false negatives (high-risk loans misclassified as healthy loans) is very important, making SVM the superior choice for this task.